src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -301,30 +322,39 @@ private void TryClosingDocumentsForMoniker(string moniker)
 
                 UnsubscribeFromWatchedHierarchies(moniker);
 
-                _workspace.ApplyChangeToWorkspace(w =>
+                // Don't do expensive workspace work on the UI thread.  Queue this up to be processed in the BG.
+                _closeDocumentQueue.AddWork(moniker);
+            }
+
+            private ValueTask ProcessCloseDocumentWorkspaceActionsAsync(ImmutableArray<string> monikers, CancellationToken _)
+            {
+                return _workspace.ApplyChangeToWorkspaceMaybeAsync(useAsync: true, w =>
                 {
-                    var documentIds = w.CurrentSolution.GetDocumentIdsWithFilePath(moniker);
-                    if (documentIds.IsDefaultOrEmpty)
+                    foreach (var moniker in monikers)
```

Comment:
i could also lift this loop *outside* of the _workspace.apply call.  I'm not sure which is preferred.  Taht would break things up into lots of little pieces of work, but would acquire/release the lock a lot more.  I figure throughput is desirable here, so we jsut take the lock once and then do all of our processing at once.